From a04860eccc6edfb71ca261d9bc242ec6590af3ba Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Wed, 19 Apr 2017 13:44:21 +0800
Subject: [PATCH 1/2] usb: gadget: udc: snps_udc: Delete repeated calling
 function

This is to fix below calltrace:

unble to handle kernel NULL pointer dereference at virtual address 00000090
pgd = ffffffc1f0582000
[00000090] *pgd=0000000270583003, *pud=0000000270583003, *pmd=0000000000000000
Internal error: Oops: 96000006 [#1] PREEMPT SMP
CPU: 1 PID: 864 Comm: modprobe Not tainted 4.8.20-rt10-WR9.0.0.5_preempt-rt #4
Hardware name: Broadcom NS2 SVK (DT)
task: ffffffc1f0604100 task.stack: ffffffc1f6620000
PC is at composite_disconnect+0x3c/0x108 [libcomposite]
LR is at composite_disconnect+0x3c/0x108 [libcomposite]
pc : [<ffffff80008bf4ec>] lr : [<ffffff80008bf4ec>] pstate: 60000145
sp : ffffffc1f6623d20
[<ffffff80008bf4ec>] composite_disconnect+0x3c/0x108 [libcomposite]
[<ffffff800080f33c>] stop_udc+0x13c/0x150 [snps_udc]
[<ffffff800080f384>] snps_gadget_stop+0x34/0x90 [snps_udc]
[<ffffff80007fb4b8>] usb_gadget_remove_driver+0x58/0x78 [udc_core]
[<ffffff80007fb5a0>] usb_gadget_unregister_driver+0xc8/0x108 [udc_core]
[<ffffff80008bf730>] usb_composite_unregister+0x20/0x30 [libcomposite]
[<ffffff80008e7330>] msg_cleanup+0x30/0x48 [g_mass_storage]
[<ffffff800811c734>] SyS_delete_module+0x1b4/0x1f8
[<ffffff8008082f30>] el0_svc_naked+0x24/0x28

The reason is that the function stop_udc called twice when remove the
module g_mass_storage and caused NULL pointer of gadget->dev, the call
relationship:

usb_gadget_remove_driver
	usb_gadget_disconnect
		snps_gadget_pullup
			stop_udc
	usb_gadget_udc_stop
		snps_gadget_stop
			stop_udc

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/usb/gadget/udc/snps_udc.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/drivers/usb/gadget/udc/snps_udc.c b/drivers/usb/gadget/udc/snps_udc.c
index 25eeafc..97f25f1 100644
--- a/drivers/usb/gadget/udc/snps_udc.c
+++ b/drivers/usb/gadget/udc/snps_udc.c
@@ -1444,7 +1444,6 @@ static int snps_gadget_stop(struct usb_gadget *gadget)
 	udc = container_of(gadget, struct snps_udc, gadget);
 
 	spin_lock_irqsave(&udc->lock, flags);
-	stop_udc(udc);
 	udc->gadget.dev.driver = NULL;
 	udc->gadget_driver = NULL;
 
-- 
1.7.5.4

